src/Microsoft.Performance.SDK/Processing/DataSourceAttribute.cs

```diff
@@ -6,32 +6,158 @@
+        /// <exception cref="System.ArgumentException">
+        ///     <paramref name="description"/> is whitespace.
+        /// </exception>
+        /// <exception cref="System.ArgumentNullException">
+        ///     <paramref name="description"/> is null.
+        /// </exception>
+        protected DataSourceAttribute(Type type, string description)
+        {
+            Guard.NotNull(type, nameof(type));
+            Guard.NotNullOrWhiteSpace(description, nameof(description));
+
+            if (!type.Is<IDataSource>())
+            {
+                throw new ArgumentException($"'{type}' must implement IDataSoure.", nameof(type));
+            }
+
+            this.Type = type;
+            this.Description = description;
+        }
+
+        /// <summary>
+        ///     Gets the <see cref="Type"/> of Data Source denoted by
+        ///     this <see cref="DataSourceAttribute"/>.
+        /// </summary>
+        public Type Type { get; }
+
+        /// <summary>
+        ///     Gets the description of the Data Source.
+        /// </summary>
+        public string Description { get; }
+
+        /// <summary>
+        ///     When overridden in a derived class, this method is used to filter any
+        ///     incoming Data Sources so that only Data Sources that could conceivably
+        ///     be processed by the decorated Custom Data Source are passed to the
+        ///     Custom Data Source for evaluation.
+        ///     <para />
+        ///     This method returns <c>true</c> if it is not overridden. This method should
+        ///     be overridden if your Data Source is potentially useful to lots of different
+        ///     Custom Data Sources (e.g. a file) but you know for a fact that most Custom
+        ///     Data Sources would not be able to do anything with them, and so you want to
+        ///     filter a subset of said Data Sources to your Custom Data Source.
+        ///     <para/>
+        ///     For example, The <see cref="FileDataSourceAttribute" /> uses this method to
+        ///     reject any <see cref="FileDataSource"/>s that do not have the prescribed extension.
+        ///     This way, only files with the relevent extension get passed to the Custom Data Source
+        ///     for further inspection. See the <see cref="FileDataSourceAttribute"/> class for
+        ///     the implementation.
+        ///     <para />
+        ///     It is guaranteed that the runtime will only pass instances of <paramref name="dataSource"/> that
+        ///     are of the <see cref="Type"/> specified by <see cref="DataSourceAttribute.Type"/>.
+        /// </summary>
+        /// <param name="dataSource">
+        ///     The Data Source to check.
+        /// </param>
+        /// <returns>
+        ///     <c>true</c> if it is feasible for this Data Source to be supported;
+        ///     <c>false</c> otherwise.
+        /// </returns>
+        public virtual bool Accepts(IDataSource dataSource)
```

Comment:
>dataSource [](start = 48, length = 10)

should we add the exception for a null parameter in the doc comments here, or do we need to do it for each implementation?

Reply:
on that note, should we implement the check here, and then call AcceptsCore(DataSource)?

---
In reply to: [579495206](https://github.com/microsoft/microsoft-performance-toolkit-sdk/pull/39#discussion_r579495206) [](ancestors = 579495206)